### ISSUE TITLE
Conversion: Add proper NUX checkout step

### ISSUE THREAD

**Issue by apeatling** [Automattic/wp-calypso#6957]
The way that checkout works in the NUX right now is a little strange. Since the checkout process is actually after you have finished the NUX flow and are logged in, it's not in the flow at all.

This means you hit the "Continue" button like you're all done, and then get dropped into the checkout flow with no instructional information at all. You aren't even able at this point to change your mind and select a different plan. If you hit the "X" then you're dumped on the plans page with no instructions.

I propose adding a checkout NUX step that within the regular NUX flow. It should be right after the selection of a plan, and allow the user to go back and change their selection.

**Checkout page before:**

<img width="1205" alt="screen shot 2016-07-21 at 13 35 29" src="https://cloud.githubusercontent.com/assets/1464705/17040038/6dfe2bac-4f52-11e6-9c59-9a3ef9d2b589.png">

**Checkout page as a NUX step (updated: v2):**

![payment-nux-step](https://cloud.githubusercontent.com/assets/1464705/17065578/7393c11c-4ff7-11e6-9b82-3f58d91305ec.png)

**Comment by apeatling** [comment]
Going to keep tweaking this, need to take into account some other things that can go in the cart in NUX (google apps etc). Also I think showing the monthly price here so big is probably confusing at a glance.

**Comment by folletto** [comment]
This is a solid idea. My feedback at this early stage would be mostly in terms of full flow. Checkout is a complex process, so we need to have the design for the full branching journey before starting building it (what's the next step? how do we handle receipts and error messages in NUX? etc).

Smaller details specifically on that first screen:

1. The `← Back` button should be centered as is in every other step for consistency.
2. I'd try to move the info text below, and make the button wider and clearer.
3. Having the summary on the side is nice, and has all the right bits imho to be effective. I wonder if it would be possible however to move it to the top, so one reads "plan, price, pay details, pay" linearly. The only downside would be that it would push the form down a bit, but if the first part at least stays above the fold it should work still better.
4. No PayPal option anymore? (it's in the first screen)

**Comment by apeatling** [comment]
Thanks for the feedback @folletto. One of the main goals here is to keep the primary action pay button above the 660px "fold", so it's not off the screen for more than half of our users. My first thought was to have a top version of the sidebar, but it's going to drop the button below.

I've made a few updates here, I'm going to have a think about the flows before and after. I think ideally we remove all the complexity of the domain registration details out of this particular step. This step would concentrate on converting users to paid in the most simple way possible, and offload more intense information gathering to later.

![payment-nux-step](https://cloud.githubusercontent.com/assets/1464705/17065578/7393c11c-4ff7-11e6-9b82-3f58d91305ec.png)

**Comment by folletto** [comment]
Ok, all clear 👍

**Comment by apeatling** [comment]
Here's a simple first flow, select plan, pay for plan, get notified of successful payment and then create account.

![payment-nux-flow](https://cloud.githubusercontent.com/assets/1464705/17067233/b709de24-4fff-11e6-99ae-523fcb3277b4.png)

**Comment by apeatling** [comment]
I feel like creating an account should happen before any of the payment details, which means it would need to happen very early in the NUX, before selecting a plan, which then means before a domain.

**Comment by folletto** [comment]
The flow seems good. So it would be already shippable for me. 

I can throw in a few discussion / review points, just in case they are worth it:

1. I'd try to keep the same Plan header on the selection and payment steps, and avoid "Cart". Cart is a construct that is useful in some online shopping, but for us, and especially in this scenario, doesn't add any valuable information to the user. I'd just show: Plan header, extra stuff, total.
2. I'd move the disclaimer text below the "Pay" button, IF it's legally ok to do so as I expect. 
3. I'd make "Or use PayPal" a link (underlined?) or something that feels more clickable. The visual weight is ok tho, so it's a balance there.
4. It's a good point on the creation of the account. The CC validation is already a good enough account verification method, but I agree the email is important and we could make it happen earlier. At the same time I agree with the flow above that ties together better selection + payment.

Quick mock:

![step-5-pay](https://cloud.githubusercontent.com/assets/4389/17098970/37a6bd18-525d-11e6-9be5-18188828a8dd.png)


Note: I keep surfacing that we should go away from _username_ everywhere, especially from signup (but editable later). ;)

**Comment by rralian** [comment]
IMHO, one important change we could do is not just the design, but in the workflow, I'd really like to put payment in front of the "Give us one minute and we'll move right along" step. I believe this setup time may have to do with tri-force, and creating the site with all of the default settings. I suspect we could squeak out a worthwhile bump in payment conversion by doing so.

![wait one minute step](https://cloudup.com/czr_dNyYuBp+)

**Comment by apeatling** [comment]
@rralian The above flow I posted would all be before this step.

**Comment by lamosty** [comment]
I like the new design @apeatling and @folletto. I'm trying to come up with smaller tasks/steps how to tackle this and have stumbled upon a few (implementation) questions:

1. Should we alter the original checkout components (so it will also apply to /checkout and everywhere else where we use it) or let it as it is for current checkout screens and add new components to NUX (while reusing as much as possible from the current ones)?

2. No mockups for `PayPalPaymentForm`, `CreditsPaymentBox` - not that important as we could change these according to the new credit card payment mockups but would help ([PayPal form screenshot](https://cloud.githubusercontent.com/assets/4988512/17131577/6592cd2c-531e-11e6-8df4-09db8a142957.png), [CreditsPaymentBox screenshot](https://cloud.githubusercontent.com/assets/4988512/17131588/74e11dba-531e-11e6-81e7-4d4fe822897e.png))

3. What about `FreeTrialConfirmationBox` and `FreeCartPaymentBox` - when are they used? Should we take these cases into consideration too or is `CreditCardPaymentBox`, `PaypalPaymentBox` and `CreditsPaymentBox` enough?

4. Do we want to reuse the components which can be reused or write them from scratch as a part of refactoring?

## Tasks

I/we will put the tasks below as new issues but before we do so, I'd like to get your opinions on them, whether they are any good, can be merged into one task or will require more sub-tasks:

1. Either we reuse (if possible) or write new `checkout/secure-payment-form`. It renders all the possible payment methods, decides which one to show/hide and also renders placeholders if data is loading.

2. Redesign/create new `checkout/payment-box` according to the mockups. It renders the "Secure Payment" title and then wraps the credit card details (takes child components).

3. Redesign/create new `checkout/credit-card-payment-box` and all the other "boxes".

4. Create new NUX step and add it to flow.

5. Put it behind a feature flag.

6. Add "Payment Successful" notification to the Create your Account step if the payment was successful. Probably also add payment unsuccessful so the user can try again (don't have mockups for this). [Screenshot](https://cloud.githubusercontent.com/assets/4988512/17132516/695eb9ee-5322-11e6-82d2-7b515eab7364.png).


The issue I see here is that currently, the component hierarchy is not suitable for our new checkout designs. For example, the "Secure Payment" title is rendered in different component than "All fields are required" but in our new designs, they are on the same line so we'll probably have to create brand new components just for NUX while reusing only the essentials such as `CreditCardForm` (which renders the CC input fields) but that too will have to be modified with at least CSS.

It seems that it will be quite a lot of work from the first looks. Will add more info when I find any.

**Comment by apeatling** [comment]
I would consider this step as something separate from the main checkout form. Although, I think it should try and reuse as many components as possible, while keeping the scope small.

The old checkout form can still be used for registered users when purchasing outside of the NUX flow.

A couple of other complexities that will need to be considered:

* We are not gathering domain registration details in this step, so that needs to be handled once the user exits the NUX flow. This could be a notice asking the user to fill these in, with a link to settings/form page (this is assuming the domain can still work without these details up front).
* There is no user account at this point where the NUX payment step is currently inserted. It's possible a lot of the checkout code currently assumes an account exists, although work has been done to address scenario.

**Comment by lamosty** [comment]
> I would consider this step as something separate from the main checkout form.

Alright, so will create brand new components for this just for NUX.

>  Although, I think it should try and reuse as many components as possible, while keeping the scope small.

Definitely. However, from what I've seen when analyzing the code, not much can be reused. Will see about this more.

**Comment by apeatling** [comment]
It's unfortunate we are building all of these components that can't be reused at all. It might be nice to understand the reasons why this happens, and in the future try and avoid those reasons. Might be a good topic for a Calypso hangout. /cc @mtias

**Comment by lamosty** [comment]
I split the task to smaller issues and put them into Plans: Maintenance milestone (which is probably not correct but I don't feel like creating new milestones. Could somebody else create a new one?). Hopefully the tasks are any good.

Another question: NUX steps are the same for both new users (with no account) and for current users who want to create a new site. Will this apply only to new users without an account? If yes, we don't have to recreate the `CreditsPaymentBox` because (I assume) new users don't have any credits.

**Comment by mtias** [comment]
@lamosty why do you think they can't be reused? It'd be good to have the reasoning before building new components that will just coexist with the others and will probably never converge.

**Comment by mtias** [comment]
For instance, I believe this popover content uses cart-item:

![image](https://cloud.githubusercontent.com/assets/548849/17182243/27277cc6-5423-11e6-9542-658ba77afd41.png)

Which is reused here:

![image](https://cloud.githubusercontent.com/assets/548849/17182264/339b1814-5423-11e6-9d02-0b95378262ac.png)

And both match conceptually what is shown here:

![image](https://cloud.githubusercontent.com/assets/548849/17182301/4edc710e-5423-11e6-9df9-ced9c6b8366f.png)

So the work should probably focus on updating the cart-item component with the design refresh, clean up the code while at it as necessary, and have a consistent result applied through our UI instead of a different component.

Of course, you won't be refactoring everything under the existing Cart, but isolating the pieces that are doable is how we can make steady progress.

**Comment by mtias** [comment]
The Cart Total sub-component should also converge:

![image](https://cloud.githubusercontent.com/assets/548849/17182595/4e40be98-5424-11e6-9ac9-c0c25f845e49.png)

![image](https://cloud.githubusercontent.com/assets/548849/17182598/53b2b7f0-5424-11e6-8f2c-6b7009b446b3.png)

![image](https://cloud.githubusercontent.com/assets/548849/17182612/63795a54-5424-11e6-80e5-5733eaab08b4.png)

See: https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/upgrades/cart/cart-total.jsx

**Comment by lamosty** [comment]
@mtias I didn't say you can't reuse CartItems, CartTotal and CartCoupon. I've written (among other things) that:

> Add the left column cart with plan name and items in cart. This will probably need to be a brand new component but hopefully we can reuse CartItems, CartTotal and CartCoupon.

Parts of the code can be reused but many not. Anyways, as soon as we start working on these things, we'll see exactly what can be reused and what not.

**Comment by mtias** [comment]
I was following from <a href="https://github.com/Automattic/wp-calypso/issues/6957#issuecomment-235342670">your last comment</a> about creating brand new components for NUX. Again, I'm curious to see what can't be reused and why, so we can correct that.

**Comment by lamosty** [comment]
I see. Initially, I asked @apeatling whether we should use the same checkout page/components but just refactor/redesign them or whether to create new component(s) for the NUX checkout step for which he replied that we should keep the current checkout page as it is and create new NUX checkout step components.

Could you also please read my [other comment](https://github.com/Automattic/wp-calypso/issues/6957#issuecomment-235210779) here? I've never worked with checkout code before so did a quick analysis and wrote my findings in that comment.

**Comment by meremagee** [comment]
Wow, lots of changes here. Ideally, we'd start with reskinning the checkout page to better match various flows before we change step order, functionality, and huge dependencies. If we're thinking of moving domain registration out of the signup flow and refactoring purchasing to not necessitate a user, then we need to be coordinating on a much deeper scale with Delta & Spectrum as this will have significant ripple effects in both teams' work.

**Comment by apeatling** [comment]
@lamosty I think @meremagee has a good point here. A good first step might be to introduce some of these design changes to the existing checkout page and checkout step in NUX. If the components can actually be shared, then we are also setting the stage for moving the checkout step in NUX as a future step. Doing it this way means we can test/launch improvements sooner.

I'm happy to talk this through with you and figure out the scope and setup.

**Comment by lamosty** [comment]
Good points. So if I understand this correctly, this is the order of things to do:

1. Apply design changes, one by one, to the checkout page as it is now;
2. Use the components from checkout page for NUX checkout step (and create new ones if can't reuse them);

**Comment by rralian** [comment]
We're exploring one small step of this for now, how difficult it would be to avoid the "wait" step and let that background work happen while the user is on the checkout screen. That's being tracked here. https://github.com/Automattic/wp-calypso/issues/7497

**Comment by rralian** [comment]
I'm moving this to `Plans: Maintenance` since it isn't really related to free user conversion. Also removing @lamosty as this isn't an active project. Sounds like it might actually be superseded by some efforts in Delta. /cc @meremagee

**Comment by alisterscott** [comment]
There hasn't been any activity on this issue in nearly a year and the NUX process has changed quite a lot in that time. I will close this issue for now as it's been low priority for that time, we can re-open should this be considered important to tackle at some point.
